xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,451 @@
+        buffer: Buffer { contents: u32:0xDEAD, length: u32:16 }
+    });
+    assert_eq(data1, u16:0xBEEF);
+
+    let (result2, data2) = buffer_sized_pop_checked<u32:32, u32:16>(result1.buffer);
+    assert_eq(result2, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0, length: u32:0 }
+    });
+    assert_eq(data2, u16:0xDEAD);
+
+    let (result3, data3) = buffer_sized_pop_checked<u32:32, u32:16>(result2.buffer);
+    assert_eq(result3, BufferResult {
+        status: BufferStatus::NO_ENOUGH_DATA,
+        buffer: result2.buffer
+    });
+    assert_eq(data3, u16:0);
+}
+
+// Returns `length` amount of data from a `buffer`.
+// It will fail if the buffer has no sufficient amount of data.
+// For calls that need better error handling, check `buffer_peek_checked`.
+pub fn buffer_peek<BSIZE: u32>(buffer: Buffer<BSIZE>, length: u32) -> bits[BSIZE] {
+    if buffer_has_at_least(buffer, length) == false {
+        trace_fmt!("Not enough data in the buffer!");
+        fail!("not_enough_data", ());
+    } else {
+        ()
+    };
+
+    let mask = (bits[BSIZE]:1 << length) - bits[BSIZE]:1;
+    buffer.contents & mask
+}
+
+#[test]
+fn test_buffer_peek() {
+    let buffer = Buffer { contents: u32:0xDEADBEEF, length: u32:32 };
+    assert_eq(buffer_peek(buffer, u32:0), u32:0);
+    assert_eq(buffer_peek(buffer, u32:16), u32:0xBEEF);
+    assert_eq(buffer_peek(buffer, u32:32), u32:0xDEADBEEF);
+}
+
+// Returns `length` amount of data from a `buffer` if possible and the status of the operation.
+pub fn buffer_peek_checked<BSIZE: u32> (buffer: Buffer<BSIZE>, length: u32) -> (BufferStatus, bits[BSIZE]) {
+    if buffer_has_at_least(buffer, length) == false {
+        (BufferStatus::NO_ENOUGH_DATA, bits[BSIZE]:0)
+    } else {
+        let mask = (bits[BSIZE]:1 << length) - bits[BSIZE]:1;
+        (BufferStatus::OK, buffer.contents & mask)
+    }
+}
+
+#[test]
+fn test_buffer_peek_checked() {
+    let buffer = Buffer { contents: u32:0xDEADBEEF, length: u32:32 };
+
+    let (status1, data1) = buffer_peek_checked(buffer, u32:0);
+    assert_eq(status1, BufferStatus::OK);
+    assert_eq(data1, u32:0);
+
+    let (status2, data2) = buffer_peek_checked(buffer, u32:16);
+    assert_eq(status2, BufferStatus::OK);
+    assert_eq(data2, u32:0xBEEF);
+
+    let (status3, data3) = buffer_peek_checked(buffer, u32:32);
+    assert_eq(status3, BufferStatus::OK);
+    assert_eq(data3, u32:0xDEADBEEF);
+
+    let (status4, data4) = buffer_peek_checked(buffer, u32:64);
+    assert_eq(status4, BufferStatus::NO_ENOUGH_DATA);
+    assert_eq(data4, u32:0);
+}
+
+// Creates a new buffer
+pub fn buffer_new<BSIZE: u32>() -> Buffer<BSIZE> {
+    Buffer { contents: bits[BSIZE]:0, length: u32:0 }
+}
+
+#[test]
+fn test_buffer_new() {
+    assert_eq(buffer_new<u32:32>(), Buffer { contents: u32:0, length: u32:0 });
+}
+
+// Example use case
+//
+// The Buffer structure can be used to collect enough data that is required
```

Comment:
maybe something to indicate that this apply to "proc" in the doc string? and refer to the concept of "state" and "next" function

Reply:
I provided more elaborate description